File: aufgaben/a2/src/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

int sp = 0;
int stack[MAXITEMS];

void print_stack(void) {
    printf("\n\tStack\n");
    printf(".-------+-------.\n");
    for (int i = sp; i >= 0; i--) {
        if (i == sp) {
            printf("|sp->%3d|%7d|\n", i, stack[i]);
        } else {
            printf("|%7d|%7d|\n", i, stack[i]);
        }
    }
    printf("'-------+-------'\n\n");
}

void push(int immediate) {
    if (sp > MAXITEMS) {
        printf("Stack overflow: Stack is full, not more than %d items allowed\n", MAXITEMS);
        exit(1);
    }
    stack[sp] = immediate;
    sp++;
}

int pop(void) {
    if ((sp == 0) && stack[sp] == 0) {
        printf("Stack underflow: popped from empty stack\n");
        exit(1);
    }
    sp--;
    int tmp = stack[sp];
    stack[sp] = 0;
    return tmp;
}

File: src/njvm/memory/stack.rs
use std::process::exit;

pub const MAXITEMS: u8 = 100;

#[derive(Debug)]
pub struct Stack {
    sp: u32,
    memory: [u32; MAXITEMS as usize],
}

impl Stack {
    pub fn new() -> Self {
        Stack {
            sp: 0,
            memory: [0; MAXITEMS as usize],
        }
    }
    pub fn push(&mut self, immediate: u32) {
        if self.sp > MAXITEMS as u32 {
            println!("Stack overflow: Stack is full, not more than {MAXITEMS} allowed");
            exit(1)
        }
        self.memory[self.sp as usize] = immediate;
        self.sp = self.sp + 1;
    }
    pub fn pop(&mut self) -> u32 {
        if self.sp == 0 && self.memory[self.sp as usize] == 0 {
            println!("Stack underflow: popped from empty stack");
            exit(1)
        }
        self.sp = self.sp - 1;
        let tmp = self.memory[self.sp as usize];
        self.memory[self.sp as usize] = 0;
        tmp
    }
    pub fn print_stack(&self) {
        for stack_item in (0..self.sp).rev() {
            if stack_item == self.sp {
                println!("{self:#?}")
            } else {
                println!("{self:#?}")
            }
        }
    }
}
